docs/v1/P4Runtime-Spec.mdk

```diff
@@ -6598,4 +6669,53 @@ clients are aware that the server is using a larger maximum receive message
 size. The gRPC server running the P4Runtime service must not set the maximum
 receive message size to a value smaller than the default (4MB).
 
+## P4Runtime Entries files { #sec-entries-files }
+
+The open source P4 compiler `p4c` [@p4c] implements an option to
+generate an "entries file", &ie; a file that contains all table
+entries declared via the `entries` table property within the program.
+
+An example P4~16~ program that can be used to demonstrate this
+capability is `table-entries-ternary-bmv2.p4` [@p4cTestProgramForConstEntries]:
+
+    git clone https://github.com/p4lang/p4c
+    cd p4c/testdata/p4_16_samples
+    mkdir tmp
+    p4test --arch v1model \
+        --p4runtime-files tmp/p4info.txt \
+        --p4runtime-entries-files tmp/entries.txt \
+        table-entries-ternary-bmv2.p4
+
+You can replace the `.txt` suffix of the file name `tmp/entries.txt`
+in the example command above with `.json` or `.bin`.  The `.bin`
+format is a binary P4Runtime API protobuf message format.  The `.txt`
+format is the text encoding of the same Protobuf messages.
+
+Target devices are *not* required to use this file.  For example, if a
+target has a P4 compiler back end that encodes all of the necessary
+details from the P4 source program, including the `entries` of tables,
+in a target-specific binary format, then that target might have no
+reason to generate these entries files.
+
+Some target devices might choose to generate entries files, and also
+to require doing so in order to have a correct implementation.  For
+example, a target runtime implementation might take a target-specific
+binary format for the compiled P4 program that does *not* contain any
+data describing the `entries` of tables, plus the entries file
+generated by p4c, and use the entries file to load the initial entries
+of tables into the device.
+
+The format of the entries file is a sequence of 0 or more `Update`
+messages.  All of them have `type` equal to `INSERT`, and `entity` is
```

Comment:
I suspect just a `WriteRequest`?
The protobuf parsers don't really support sequences of messages out of the box AFAIK, so if it was really just a concatenation of parsable protobuf blobs that would make things pretty hard to consume.

Reply:
Here is an example entries file written by p4c in text format: https://github.com/p4lang/p4c/blob/main/testdata/p4_16_samples_outputs/table-entries-ternary-bmv2.p4.entries.txt

It doesn't look like a `WriteRequest` message to me, since there are no fields of a WriteRequest message present there, only a sequence of 0 or more `Update` messages.  If you tell me that is a valid `WriteRequest` message, I am happy to document it that way in the P4Runtime spec.

Reply:
Yup, the link you provided does show a valid `WriteRequest`.

> It doesn't look like a WriteRequest message to me, since there are no fields of a WriteRequest message present there

Actually it does: the `repeated updates` field is a `WriteRequest` field [[1]], and the textproto you linked to contains a bunch of `updates { ... }`.

This is admittedly subtle, because the text proto format doesn't include an outer wrapper for the message that is being encoded. Instead, it just shows the fields of the message at the top-level.

That being said, the text format is ambiguous, this could be a dump of *any* proto message with a `repeated Update` field. And for the binary format, this distinction matters. So we shouldn't blindly claim this is a `WriteRequest`. Do you happen to know where in the p4c source code this output gets created? From there it will be straightforward to tell the correct message type.

[1]: http://google3/third_party/p4lang_p4runtime/proto/p4/v1/p4runtime.proto;l=73;rcl=534134847

Reply:
Here is the p4c code that writes the entries file: https://github.com/p4lang/p4c/blob/main/control-plane/p4RuntimeSerializer.cpp#L973-L976

in case the line numbers go askew, the method name is `P4RuntimeEntriesConverter`

Reply:
If you would like to examing the binary and/or JSON format of the entries file generated by current open source p4c, in addition to the text format, they are all in the zip file attached to this comment.
[tmp.zip](https://github.com/p4lang/p4runtime/files/12086082/tmp.zip)

Reply:
I didn't find the place where things get serialized, but I would consider the following a good-enough hint that the message is a `p4.v1.WriteRequest`: https://github.com/p4lang/p4c/blob/main/control-plane/p4RuntimeSerializer.cpp#L1327-L1328

I also opened https://github.com/p4lang/p4c/issues/4070 so people don't have to guess in the future.

Reply:
Is there an easy way for you to parse the contents of the binary format entries file in the file `tmp.zip` attached in my previous comment, to determine whether it is a valid binary `WriteRequest` protobuf message?  I suspect it probably is, but do not know any quick way to check.

Reply:
Confirmed with a Google-internal CLI tool that entries.bin can be parsed as a `p4.v1.WriteRequest`.